nengo/synapses.py

```diff
@@ -221,123 +223,130 @@ def evaluate(self, frequencies):
-        def __init__(self, num, den, output):
-            self.num = num
-            self.den = den
-            self.output = output
+        def __init__(self, A, B, C, D, X):
+            if not self.check(A, B, C, D, X):
+                raise ValidationError(
+                    "Matrices do not meet the requirements for this Step",
+                    attr='matrices', obj=self)
+            self.A = A
+            self.B = B
+            self.C = C
+            self.D = D
+            self.X = X
 
         def __call__(self, t, signal):
             raise NotImplementedError("Step functions must implement __call__")
 
-    class NoDen(Step):
-        """An LTI step function for transfer functions with no denominator.
-
-        This step function should be much faster than the equivalent general
-        step function.
-        """
-        def __init__(self, num, den, output):
-            if len(den) > 0:
-                raise ValidationError("'den' must be empty (got length %d)"
-                                      % len(den), attr='den', obj=self)
-            super().__init__(num, den, output)
-            self.b = num[0]
+        @classmethod
+        def check(cls, A, B, C, D, X):
+            if A.shape[0] == 0:
+                return X.shape[0] == B.shape[0] == 0 and D.shape[0] == 1
+            else:
+                return (A.shape[0] == A.shape[1] == B.shape[0] == C.shape[1]
+                        and D.shape[0] == B.shape[1] == 1
+                        and A.shape[0] == X.shape[0])
+
+    class NoX(Step):
+        """Step for transfer function with no state and single numerator."""
+        def __init__(self, A, B, C, D, X):
+            super().__init__(A, B, C, D, X)
+            self.d = D.item()
 
         def __call__(self, t, signal):
-            self.output[...] = self.b * signal
-            return self.output
+            return self.d * signal
 
-    class Simple(Step):
-        """An LTI step function for transfer functions with one num and den.
+        @classmethod
+        def check(cls, A, B, C, D, X):
+            return super(LinearFilter.NoX, cls).check(A, B, C, D, X) and (
+                len(A) == 0)
 
-        This step function should be much faster than the equivalent general
-        step function.
-        """
-        def __init__(self, num, den, output, y0=None):
-            if len(num) != 1:
-                raise ValidationError("'num' must be length 1 (got %d)"
-                                      % len(num), attr='num', obj=self)
-            if len(den) != 1:
-                raise ValidationError("'den' must be length 1 (got %d)"
-                                      % len(den), attr='den', obj=self)
-
-            super().__init__(num, den, output)
-            self.b = num[0]
-            self.a = den[0]
-            if y0 is not None:
-                self.output[...] = y0
+    class OneX(Step):
+        """Step function for transfer functions with one state element."""
```

Comment:
Also no passthrough (note you can have a one-dimensional state with a passthrough, but due to the `check`s below that will fall into the `General` case).

Reply:
Fixed